pgs.hpp

```diff
@@ -0,0 +1,292 @@
+        pcs.compositionObjects[i].verticalPosition              = swapEndianness(*(uint16_t*)&buffer[bufferStartIdx + 6]);
+        if (pcs.compositionObjects[i].croppedAndForcedFlag & e_objectFlags::cropped) {
+            pcs.compositionObjects[i].croppedHorizontalPosition = swapEndianness(*(uint16_t*)&buffer[bufferStartIdx + 8]);
+            pcs.compositionObjects[i].croppedVerticalPosition   = swapEndianness(*(uint16_t*)&buffer[bufferStartIdx + 10]);
+            pcs.compositionObjects[i].croppedWidth              = swapEndianness(*(uint16_t*)&buffer[bufferStartIdx + 12]);
+            pcs.compositionObjects[i].croppedHeight             = swapEndianness(*(uint16_t*)&buffer[bufferStartIdx + 14]);
+            bufferStartIdx += 8;
+        }
+        bufferStartIdx += 8;
+    }
+
+    return pcs;
+}
+
+void t_PCS::write(uint8_t* buffer) {
+    *((uint16_t*)(&buffer[0]))  = swapEndianness(width);
+    *((uint16_t*)(&buffer[2]))  = swapEndianness(height);
+    *((uint8_t*) (&buffer[4]))  =                frameRate;
+    *((uint16_t*)(&buffer[5]))  = swapEndianness(compositionNumber);
+    *((uint8_t*) (&buffer[7]))  =                compositionState;
+    *((uint8_t*) (&buffer[8]))  =                paletteUpdateFlag;
+    *((uint8_t*) (&buffer[9]))  =                paletteID;
+    *((uint8_t*) (&buffer[10])) =                numberOfCompositionObjects;
+
+    size_t bufferStartIdx = 11;
+    for (int i = 0; i < numberOfCompositionObjects; i++) {
+        *((uint16_t*)(&buffer[bufferStartIdx + 0]))  = swapEndianness(compositionObjects[i].objectID);
+        *((uint8_t*) (&buffer[bufferStartIdx + 2]))  =                compositionObjects[i].windowID;
+        *((uint8_t*) (&buffer[bufferStartIdx + 3]))  =                compositionObjects[i].croppedAndForcedFlag;
+        *((uint16_t*)(&buffer[bufferStartIdx + 4]))  = swapEndianness(compositionObjects[i].horizontalPosition);
+        *((uint16_t*)(&buffer[bufferStartIdx + 6]))  = swapEndianness(compositionObjects[i].verticalPosition);
+        if (compositionObjects[i].croppedAndForcedFlag & e_objectFlags::cropped) {
+            *((uint16_t*)(&buffer[bufferStartIdx + 8]))  = swapEndianness(compositionObjects[i].croppedHorizontalPosition);
+            *((uint16_t*)(&buffer[bufferStartIdx + 10])) = swapEndianness(compositionObjects[i].croppedVerticalPosition);
+            *((uint16_t*)(&buffer[bufferStartIdx + 12])) = swapEndianness(compositionObjects[i].croppedWidth);
+            *((uint16_t*)(&buffer[bufferStartIdx + 14])) = swapEndianness(compositionObjects[i].croppedHeight);
+            bufferStartIdx += 8;
+        }
+        bufferStartIdx += 8;
+    }
+}
+
+
+t_PDS t_PDS::read(uint8_t* buffer, size_t segment_size) {
+    t_PDS pds;
+
+    pds.id            = *(uint8_t*)&buffer[0];
+    pds.versionNumber = *(uint8_t*)&buffer[1];
+
+    pds.numberOfPalettes = (segment_size - 2) / 5;
+
+    for (int i = 0; i < pds.numberOfPalettes; i++) {
```

Comment:
There are bad streams that do allocate palette ID 0xFF. as pds.palettes has 255 entries, this is a segfault waiting to happen.

I would suggest to:
- Check if the entry is transparent (alpha=0) 
- If it isn't, print an error.
- In either case, not store it. As entry 0xFF is internally defined to be transparent by all PGS decoders.

[There are evidence](https://forum.doom9.org/showthread.php?t=121535&highlight=s&page=39) that using the 0xFF entry leads to catastrophic behaviour if it is not transparent.